File: Solutions/0671- Second Minimum Node In a Binary Tree/Solution.py
class Solution:
    def __init__(self):
        self.firstMin = None
        self.secondMin = -1
    
    def findSecondMinimumValue(self, root: TreeNode) -> int:
        if root is None:
            return -1
        
        self.firstMin = root.val
        self.dfs(root)
        return self.secondMin if self.secondMin != -1 else -1

    def dfs(self, root: TreeNode):
        if root is None:
            return
        
        if self.firstMin != root.val and (self.secondMin == -1 or root.val < self.secondMin):
            self.secondMin = root.val

        self.dfs(root.left)
        self.dfs(root.right)

File: Solutions/0671- Second Minimum Node In a Binary Tree/Solution.java
class Solution {
    int firstMin;
    int secondMin = -1;

    public int findSecondMinimumValue(TreeNode root) {
        firstMin = root.val;
        dfs(root);
        return secondMin;

    }

    private void dfs(TreeNode root) {

        if (root.val != firstMin && (secondMin == -1 || root.val < secondMin)) {
            secondMin = root.val;
        }

        if (root.left != null) {
            dfs(root.left);
            dfs(root.right);
        }
    }
}